Add css-individual-transform-properties blogpost (Fix #8244) (#8375)
GoogleChrome/web.dev: src/site/content/en/blog/css-individual-transform-properties/index.md

---
layout: post
title: Finer grained control over CSS transforms with individual transform properties
authors:
  - bramus
  - dbaron

> description: >

Reviewer: We need a hero image and alt text. https://web.dev/handbook/use-media/

  Learn how you can use the individual translate, rotate, and scale CSS properties to approach transforms in an intuitive way.

> date: 2022-08-02

Reviewer: ```suggestion
date: 2022-08-02
tags:
  - blog
  - newly-interoperable
```

Reviewer: It won't show up on the blog without the blog tag.

---

> 

Reviewer: Add the aside as in this post https://web.dev/transformstream/

I'll go through and tidy these up once we're not "newly interoperable" any more, as I can find them via the newly-interoperable tag.

```

The targeted element is translated by 50% on the X-axis, rotated by 30 degrees, and finally scaled up to 120%.

While the `transform` property does its work just fine, it becomes a bit tedious when you want to alter any of those values individually.

To change the scale on hover, you must duplicate all functions in the transform property, even though their values remain unchanged.

```css
.target:hover {
  transform: translateX(50%) rotate(30deg) scale(2); /* Only the value of scale() changed */
}
```

## The individual transform properties

Shipping with Chrome 104 are individual properties for CSS transforms. The properties are `scale`, `rotate`, and `translate`, which you can use to individually define those parts of a transformation.

By doing so, Chrome joins Firefox and Safari which already support these properties.

{% BrowserCompat 'css.properties.scale' %}

{% Aside %}
Not all transformation functions have a matching individual property, for example `skewX()` and `matrix()`.
{% endAside %}

Rewriting the preceding `transform` example with the individual properties, your snippet becomes this:

```css
.target {
  translate: 50% 0;
  rotate: 30deg;
  scale: 1.2;
}
```

> ## Order Matters

Reviewer: ```suggestion
## Order matters
```

One key difference between the original CSS `transform` property and the new properties is the order in which the declared transformations get applied.

With `transform`, the transformation functions get applied in the order they’re written–from left (outside) to right (inside).

With the individual transformation properties, the order is not the order in which they are declared.  The order is always the same: first `translate` (outside), then `rotate`, and then `scale` (inside).

> That means both code snippets below yield the same result:

Reviewer: ```suggestion
That means both of the following code snippets give the same result:
```

```css
.transform--individual {
  translate: 50% 0;
  rotate: 30deg;
  scale: 1.2;
}

.transform--individual-alt {
  rotate: 30deg;
  translate: 50% 0;
  scale: 1.2;
}
```

In both cases the targeted elements will first be translated by `50%` on the X-axis, then rotated by `30deg`, and finally scaled by `1.2`.

If one of the individual transform properties are declared along with a `transform` property, then the individual transforms get applied first (`translate`, `rotate`, and then `scale`) with the `transform` last (inside). More details are in the spec that defines [how the transformation matrix should be calculated](https://www.w3.org/TR/css-transforms-2/#ctm).

## Animations

The main reason these properties were added is to make animations easier. Say you want to animate an element as follows:

> {% Img src="image/AeNB0cHNDkYPUYzDuv8gInYA9rY2/Pn0t3nrqfBg6JhmYFslm.svg", alt="Keyframes Graph", width="800", height="745" %}

Reviewer: ```suggestion
{% Img src="image/AeNB0cHNDkYPUYzDuv8gInYA9rY2/Pn0t3nrqfBg6JhmYFslm.svg", alt="Keyframes graph.", width="800", height="745" %}
```

### Using `transform`

To implement this animation using `transform`, you’d have to calculate all in-between values for all defined transformations, and include those in each keyframe. For example, to do a rotation at the 10% mark, the values for the other transformations must be calculated as well, because the `transform` property needs all of them.

> {% Img src="image/AeNB0cHNDkYPUYzDuv8gInYA9rY2/AF5n6UlhcuQ5UNKemX8M.svg", alt="Keyframes Graph with intermediate values calculated", width="800", height="745" %}

Reviewer: ```suggestion
{% Img src="image/AeNB0cHNDkYPUYzDuv8gInYA9rY2/AF5n6UlhcuQ5UNKemX8M.svg", alt="Keyframes graph with intermediate values calculated.", width="800", height="745" %}
```

The resulting CSS code becomes this:

```css
@keyframes anim {
  0% { transform: translateX(0%); }
  5% { transform: translateX(5%) rotate(90deg) scale(1.2); }
  10% { transform: translateX(10%) rotate(180deg) scale(1.2); }
  90% { transform: translateX(90%) rotate(180deg) scale(1.2); }
  95% { transform: translateX(95%) rotate(270deg) scale(1.2); }
  100% { transform: translateX(100%) rotate(360deg); }
}

.target {
  animation: anim 2s;
  animation-fill-mode: forwards;
}
```

> {% Codepen {

Reviewer: I need to send you the web.dev CodePen login details, as it's better they are in there as then anyone can update them if needed. Just fork yours into the account.
